File: tree.js
const fs = require('fs');
const path = require('path');

function print_help() {
    console.log("Usage: tree [options] [directory]");
    console.log("Options:");
    console.log("  -a    Show all files (including hidden)");
    console.log("  -L    Limit depth");
    console.log("  -h    Show this help");
}

function clearParam(param) {
    if (!param) return path.resolve(".");
    if (param.includes('"')) {
        return path.resolve(param.replace(/"/g, "\\").split(' ')[0]);
    }
    return path.resolve(param.trim());
}

const args = process.argv.slice(2);

if (args.includes('-h') || args.includes('--help')) {
    print_help();
    process.exit(0);
}

const targetArg = process.argv[2] || '.';
const directoryPath = clearParam(targetArg);

if (!fs.existsSync(directoryPath)) {
    console.error(`Error: Directory '${directoryPath}' not found`);
    process.exit(1);
}

if (!fs.statSync(directoryPath).isDirectory()) {
    console.error(`Error: '${directoryPath}' is not a directory`);
    process.exit(1);
}

function tree(dir, prefix = '', isLast = true) {
    const children = fs.readdirSync(dir)
        .filter(name => !args.includes('-a') || name[0] !== '.')
        .map(name => ({ name, path: path.join(dir, name) }));
    
    children.forEach((child, index) => {
        const isLastChild = index === children.length - 1;
        const line = prefix + (isLastChild ? "└── " : "├── ") + child.name;
        console.log(line);
        
        if (fs.statSync(child.path).isDirectory()) {
            const newPrefix = prefix + (isLastChild ? "    " : "│   ");
            tree(child.path, newPrefix, isLastChild);
        }
    });
}

console.log(directoryPath);
tree(directoryPath);

File: tree.py
import os
import sys
from pathlib import Path

def print_help():
    print("Usage: tree [options] [directory]")
    print("Options:")
    print("  -a    Show all files (including hidden)")
    print("  -L    Limit depth")
    print("  -h    Show this help")

def tree(directory, prefix="", is_last=True, show_hidden=False, max_depth=None, current_depth=0):
    if max_depth is not None and current_depth >= max_depth:
        return
    
    try:
        entries = sorted(os.listdir(directory))
        
        if not show_hidden:
            entries = [e for e in entries if not e.startswith('.')]
        
        for i, entry in enumerate(entries):
            path = os.path.join(directory, entry)
            is_entry_last = (i == len(entries) - 1)
            
            if os.path.isdir(path):
                print(f"{prefix}{'└── ' if is_last else '├── '}{entry}/")
                new_prefix = prefix + ('    ' if is_last else '│   ')
                tree(path, new_prefix, is_entry_last, show_hidden, max_depth, current_depth + 1)
            else:
                print(f"{prefix}{'└── ' if is_last else '├── '}{entry}")
    except PermissionError:
        print(f"{prefix}{'└── ' if is_last else '├── '}[Permission denied]")

def main():
    args = sys.argv[1:]
    
    if not args:
        tree('.')
        return 0
    
    show_hidden = False
    max_depth = None
    directory = '.'
    
    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-a':
            show_hidden = True
        elif arg == '-L' and i + 1 < len(args):
            try:
                max_depth = int(args[i + 1])
                i += 1
            except ValueError:
                print(f"Error: Invalid depth value: {args[i + 1]}")
                return 1
        elif arg in ('-h', '--help', '/?'):
            print_help()
            return 0
        elif not arg.startswith('-'):
            directory = arg
        i += 1
    
    if not os.path.isdir(directory):
        print(f"Error: '{directory}' is not a directory")
        return 1
    
    tree(directory, '', True, show_hidden, max_depth)
    
    return 0

if __name__ == "__main__":
    sys.exit(main())